### PR TITLE
Update logging to use existing cluster deployment for defaults

### PR DESCRIPTION
**Pull request by ewolinetz**
This will allow us to use logging facts to set defaults of specific configurations such as ES index replicas and shard count.

The update to logging facts yields us output like: 
```json
"elasticsearch": {
            "clusterrolebindings": {}, 
            "configmaps": {
                "logging-elasticsearch": {
                    "elasticsearch.yml": {
                        "cloud": {
                            "kubernetes": {
                                "namespace": "${NAMESPACE}", 
                                "pod_label": "${POD_LABEL}", 
                                "pod_port": 9300
                            }
                        }, 
                        "cluster": {
                            "name": "${CLUSTER_NAME}"
                        }, 
                        "discovery": {
                            "type": "kubernetes", 
                            "zen.minimum_master_nodes": "${NODE_QUORUM}", 
                            "zen.ping.multicast.enabled": false
                        }, 
                        "gateway": {
                            "expected_nodes": "${RECOVER_EXPECTED_NODES}", 
                            "recover_after_nodes": "${NODE_QUORUM}", 
                            "recover_after_time": "${RECOVER_AFTER_TIME}"
                        }, 
                        "index": {
                            "number_of_replicas": 0, 
                            "number_of_shards": 1, 
                            "translog": {
                                "flush_threshold_period": "5m", 
                                "flush_threshold_size": "256mb"
                            }, 
                            "unassigned.node_left.delayed_timeout": "2m"
                        }, 
                        "io.fabric8.elasticsearch.authentication.users": [
                            "system.logging.kibana", 
                            "system.logging.fluentd", 
                            "system.logging.curator", 
                            "system.admin"
                        ], 
```

TODO:
- [x] Update logging facts to pull out settings from config maps
- [x] Move `openshift_sanitize_inventory/library/conditional_set_fact.py` up to repo level
- [x] Generate diffs against currently deployed configs and correctly patch in custom changes from customers
- [x] Use `conditional_set_fact` to easily set defaults for logging based on logging facts, or falling back to role defaults when not specified in the inventory
- [x] Update all components to follow patching configmaps

### CODE REVIEW

**Comment by ewolinetz** [comment]
@portante other than the ES replica and shard count, are there other ES specific settings we should ensure we reuse in a more sane way?

**Comment by ewolinetz** [comment]
logging flake
/retest

**Comment by richm** [comment]
logging is broken until we fix the issue with fluent-plugin-kubernetes_metadata_filter 0.30.0 @jcantrill

**Comment by ewolinetz** [comment]
I think something is up with the tox test... It's saying the following lines need 6 spaces instead of 4:

https://github.com/openshift/openshift-ansible/pull/5894/files#diff-f4c6ebdf7da2e4a833073d1f89e08221R66

https://github.com/openshift/openshift-ansible/pull/5894/files#diff-642c8aef5cb4343e18a86a22446da3bdR127

https://github.com/openshift/openshift-ansible/pull/5894/files#diff-f0d35e3da9e7cc37ec20ba6719fa326bR105

**Comment by ewolinetz** [comment]
flake on getting repo data?
/retest

**Comment by ewolinetz** [comment]
/retest

Failure in tox:
```
--- FAIL: TestPackageUpdateDepMissing (25.43s)
	common.go:54: missing in output: ["check \"package_update\":" "Could not perform a yum update." "break-yum-update-1.0-2.noarch requires package-that-does-not-exist"]
	common.go:98: 
		$ (cd /data/src/github.com/openshift/openshift-ansible/test/integration/openshift_health_checker/preflight && ansible-playbook -e testing_skip_some_requirements=1 -i /dev/null playbooks/package_update_dep_missing.yml)
		...
		
		
		
		Failure summary:
		
		
		  1. Hosts:    openshift_ansible_test_3379044966895, openshift_ansible_test_3379044966895
		     Play:     Initialize host facts
		     Task:     openshift_sanitize_inventory : Assign deprecated variables to correct counterparts
		     Message:  [0;31mFailed without returning a message.[0m
--- FAIL: TestPackageAvailabilityMissingRequired (26.62s)
	common.go:54: missing in output: ["check \"package_availability\":" "Cannot install all of the necessary packages." "atomic-openshift"]
	common.go:98: 
		$ (cd /data/src/github.com/openshift/openshift-ansible/test/integration/openshift_health_checker/preflight && ansible-playbook -e testing_skip_some_requirements=1 -i /dev/null playbooks/package_availability_missing_required.yml)
		...
		
		
		
		Failure summary:
		
		
		  1. Hosts:    openshift_ansible_test_31158775042260, openshift_ansible_test_31158775042260
		     Play:     Initialize host facts
		     Task:     openshift_sanitize_inventory : Assign deprecated variables to correct counterparts
		     Message:  [0;31mFailed without returning a message.[0m
--- FAIL: TestPackageVersionMismatches (27.20s)
	common.go:54: missing in output: ["check \"package_version\":" "Not all of the required packages are available at their requested version"]
	common.go:98: 
		$ (cd /data/src/github.com/openshift/openshift-ansible/test/integration/openshift_health_checker/preflight && ansible-playbook -e testing_skip_some_requirements=1 -i /dev/null playbooks/package_version_mismatches.yml)
		...
		
		
		
		Failure summary:
		
		
		  1. Hosts:    openshift_ansible_test_52725369679804, openshift_ansible_test_52725369679804
		     Play:     Initialize host facts
		     Task:     openshift_sanitize_inventory : Assign deprecated variables to correct counterparts
		     Message:  [0;31mFailed without returning a message.[0m
--- FAIL: TestPackageUpdateRepoUnreachable (28.12s)
	common.go:54: missing in output: ["check \"package_update\":" "Error getting data from at least one yum repository"]
	common.go:98: 
		$ (cd /data/src/github.com/openshift/openshift-ansible/test/integration/openshift_health_checker/preflight && ansible-playbook -e testing_skip_some_requirements=1 -i /dev/null playbooks/package_update_repo_unreachable.yml)
		...
		
		
		
		Failure summary:
		
		
		  1. Hosts:    openshift_ansible_test_81382626541916, openshift_ansible_test_81382626541916
		     Play:     Initialize host facts
		     Task:     openshift_sanitize_inventory : Assign deprecated variables to correct counterparts
		     Message:  [0;31mFailed without returning a message.[0m
--- FAIL: TestPackageAvailabilitySucceeds (25.04s)
	common.go:80: got exit code 2, want 0
	common.go:98: 
		$ (cd /data/src/github.com/openshift/openshift-ansible/test/integration/openshift_health_checker/preflight && ansible-playbook -e testing_skip_some_requirements=1 -i /dev/null playbooks/package_availability_succeeds.yml)
		...
		
		
		
		Failure summary:
		
		
		  1. Hosts:    openshift_ansible_test_41569362388210, openshift_ansible_test_41569362388210
		     Play:     Initialize host facts
		     Task:     openshift_sanitize_inventory : Assign deprecated variables to correct counterparts
		     Message:  [0;31mFailed without returning a message.[0m
--- FAIL: TestPackageVersionMatches (26.24s)
	common.go:80: got exit code 2, want 0
	common.go:98: 
		$ (cd /data/src/github.com/openshift/openshift-ansible/test/integration/openshift_health_checker/preflight && ansible-playbook -e testing_skip_some_requirements=1 -i /dev/null playbooks/package_version_matches.yml)
		...
		
		
		
		Failure summary:
		
		
		  1. Hosts:    openshift_ansible_test_77722627616023, openshift_ansible_test_77722627616023
		     Play:     Initialize host facts
		     Task:     openshift_sanitize_inventory : Assign deprecated variables to correct counterparts
		     Message:  [0;31mFailed without returning a message.[0m
--- FAIL: TestPackageVersionMultiple (26.54s)
	common.go:54: missing in output: ["check \"package_version\":" "Multiple minor versions of these packages are available"]
	common.go:98: 
		$ (cd /data/src/github.com/openshift/openshift-ansible/test/integration/openshift_health_checker/preflight && ansible-playbook -e testing_skip_some_requirements=1 -i /dev/null playbooks/package_version_multiple.yml)
		...
		
		
		
		Failure summary:
		
		
		  1. Hosts:    openshift_ansible_test_5477082571912, openshift_ansible_test_5477082571912
		     Play:     Initialize host facts
		     Task:     openshift_sanitize_inventory : Assign deprecated variables to correct counterparts
		     Message:  [0;31mFailed without returning a message.[0m
--- FAIL: TestPackageUpdateRepoDisabled (26.42s)
	common.go:80: got exit code 2, want 0
	common.go:98: 
		$ (cd /data/src/github.com/openshift/openshift-ansible/test/integration/openshift_health_checker/preflight && ansible-playbook -e testing_skip_some_requirements=1 -i /dev/null playbooks/package_update_repo_disabled.yml)
		...
		
		
		
		Failure summary:
		
		
		  1. Hosts:    openshift_ansible_test_18715211543457, openshift_ansible_test_18715211543457
		     Play:     Initialize host facts
		     Task:     openshift_sanitize_inventory : Assign deprecated variables to correct counterparts
		     Message:  [0;31mFailed without returning a message.[0m
--- FAIL: TestPackageUpdateRepoBroken (16.94s)
	common.go:54: missing in output: ["check \"package_update\":" "Error with yum repository configuration: Cannot find a valid baseurl for repo"]
	common.go:98: 
		$ (cd /data/src/github.com/openshift/openshift-ansible/test/integration/openshift_health_checker/preflight && ansible-playbook -e testing_skip_some_requirements=1 -i /dev/null playbooks/package_update_repo_broken.yml)
		...
		
		
		
		Failure summary:
		
		
		  1. Hosts:    openshift_ansible_test_61005122221938, openshift_ansible_test_61005122221938
		     Play:     Initialize host facts
		     Task:     openshift_sanitize_inventory : Assign deprecated variables to correct counterparts
		     Message:  [0;31mFailed without returning a message.[0m
FAIL
FAIL	_/data/src/github.com/openshift/openshift-ansible/test/integration/openshift_health_checker/preflight	67.422s
```

**Comment by ewolinetz** [comment]
bot, retest this please

File: roles/openshift_logging/tasks/install_logging.yaml [hunk 1]
```diff
@@ -4,6 +4,19 @@
     oc_bin: "{{openshift.common.client_binary}}"
     openshift_logging_namespace: "{{openshift_logging_namespace}}"
 
+- conditional_set_fact:
+    facts: "{{ openshift_logging_facts['elasticsearch']['configmaps']['logging-elasticsearch']['elasticsearch.yml'] | flatten_dict }}"
```
Comment (jcantrill) [inline roles/openshift_logging/tasks/install_logging.yaml]: Do we need to be concerned about the difference between an ops and non-ops deployment?

**Comment by ewolinetz** [comment]
/retest
bot, retest this please

File: roles/openshift_logging/tasks/install_logging.yaml [hunk 2]
```diff
@@ -4,6 +4,28 @@
     oc_bin: "{{openshift.common.client_binary}}"
     openshift_logging_namespace: "{{openshift_logging_namespace}}"
 
+- conditional_set_fact:
```
Comment (michaelgugino) [inline roles/openshift_logging/tasks/install_logging.yaml]: This section is not very clear what's happening to what.

File: roles/openshift_logging/tasks/build_patch.yaml [hunk 3]
```diff
@@ -0,0 +1,58 @@
+---
```
Comment (michaelgugino) [inline roles/openshift_logging/tasks/build_patch.yaml]: I have some hesitation to implement something like this.  Why is building such a complicated set of patch files necessary?

If anything, this looks like it should be created as an action plugin, and then we register the output of that plugin to create whatever file/patch we need.

As-is, this is pretty hard to follow.

File: roles/openshift_logging_fluentd/tasks/main.yaml [hunk 4]
```diff
@@ -108,38 +108,28 @@
...
-    dest: "{{ tempdir }}/fluentd-throttle-config.yaml"
-  when: fluentd_throttle_contents is defined
-  changed_when: no
-
-- copy:
-    content: "{{ fluentd_secureforward_contents }}"
-    dest: "{{ tempdir }}/secure-forward.conf"
-  when: fluentd_secureforward_contents is defined
-  changed_when: no
+- include_role:
```
Comment (michaelgugino) [inline roles/openshift_logging_fluentd/tasks/main.yaml]: This looks like it should possibly be a plugin instead of a role, based on the actions it takes.

File: roles/openshift_logging/tasks/build_patch.yaml (hunk 3 above)
Comment (ewolinetz) [reply roles/openshift_logging/tasks/build_patch.yaml]: There is a requirement to maintain any custom configuration changes to currently deployed configmap files for logging while still having a way to patch in any necessary changes that the installer may need to make to allow our images to work.

I could try to rewrite this as a plugin instead.

File: roles/openshift_logging/tasks/install_logging.yaml (hunk 2 above)
Comment (ewolinetz) [reply roles/openshift_logging/tasks/install_logging.yaml]: Would adding comments help? The goal here is to reuse variables from a currently deployed configuration as a default rather than a role default when able. So we are pulling out the value from a variable if the variable is defined using the `conditional_set_fact` module from the logging facts. Afterwards we do something similar using the inventory, using `conditional_set_fact` saves us from doing a bunch of `when: <var> is defined` for each possible variable we may do this for

File: roles/openshift_logging_fluentd/tasks/main.yaml (hunk 4 above)
Comment (ewolinetz) [reply roles/openshift_logging_fluentd/tasks/main.yaml]: I'll take a stab at that

**Comment by richm** [comment]
/test logging
curator flake

**Comment by richm** [comment]
logging - not a flake - something wrong with kibana deployment

File: roles/openshift_logging_fluentd/tasks/main.yaml (hunk 4 above)
Comment (ewolinetz) [reply roles/openshift_logging_fluentd/tasks/main.yaml]: I couldn't find a really easy way to patch files in python, that and the extra oc_module usage makes me think it is most simple to leave it as a task list. I did however convert the task which generated the patch into a module call

**Comment by ewolinetz** [comment]
@michaelgugino addressed your comments, PTAL

**Comment by michaelgugino** [comment]
@ewolinetz Yeah, I think we can just generate the patch file, and apply that, is what I'm suggesting, instead of scraping the existing config map, patching it locally, and then applying.

As for idempotency, I would leave the patch file in place somewhere on the master.  I would probably make the patch file a template that just dumps some dictionary we pass it after filtering.  Register the change, only execute if that file has been updated.

Alternatively, before and after patching, register the output of the config map and compare.  This might require writing a custom module.

Does that sound feasible?

**Comment by ewolinetz** [comment]
@michaelgugino I'm not sure if that's feasible... at least not as I'm understanding your suggestion.

So the purpose of this change is to maintain any custom changes customers may have done to their existing configmaps for logging during an upgrade. So we would need to scrape the current configmaps in order to do that and find out what changes were made (via `diff`). Ultimately we want to apply any configmap updates that the role is providing, so we would want to update what currently exists idempotently. But we also want to do our best to maintain anything custom a customer may have added. 

Before we accomplished this by letting a customer provide the contents of the file that we should use, however that is dangerous because any necessary configurations we need to provide may never land.

**Comment by michaelgugino** [comment]
The API's patch method seems to do what you're saying.  "Just add some stuff to an existing config" is how I read it.

> Notes on the strategic merge patch
With a patch, you do not have to specify an entire object; you specify only the portion of the object that you want to change. For example, in the preceding exercise, you specified one Container in the containers list in a PodSpec.
... With a strategic merge patch, you can update a list by specifying only the elements that you want to add to the list. The existing list elements remain, and the new elements are merged with the existing elements. 

So, the end user can have whatever they please in the config map, but we can just patch the portions that we deem necessary.

**Comment by ewolinetz** [comment]
@michaelgugino correct, however that would be the same amount of work or more (for configmaps that have more than one file entry in it) than to use `oc apply`. That would also cause us to need to check if the configmap already exists, and if it does we patch it, otherwise we would create it new.

The parts of the object we update would be the key/value pairs that are the data entries within the configmap. Unless I'm mistaken there is no way to update the file contents (the value) within the configmap data, we would patch to replace the contents.

File: roles/openshift_logging/filter_plugins/openshift_logging.py [hunk 5]
```diff
@@ -102,6 +102,38 @@ def serviceaccount_namespace(qualified_sa, default=None):
     return seg[-1]
 
 
+def build_protected_line_string(protected_lines):
```
Comment (richm) [inline roles/openshift_logging/filter_plugins/openshift_logging.py]: Can you give an example of where/how this is used?

File: roles/openshift_logging/tasks/patch_configmap_files.yaml [hunk 6]
```diff
@@ -0,0 +1,31 @@
+---
+## The purpose of this task file is to take in a list of configmap files provided
+##  in the variable configmap_file_names, which correspond to the data sections
+##  within a configmap. We iterate over each of these files and create a patch
+##  from the diff between current_file and new_file to try to maintain any custom
+##  changes that a customer may have made to a currently deployed configmap while
```
Comment (michaelgugino) [inline roles/openshift_logging/tasks/patch_configmap_files.yaml]: Probably 'user' would be better verbiage than 'customer' :)

File: roles/openshift_logging/library/openshift_logging_facts.py [hunk 7]
```diff
@@ -214,7 +222,10 @@ def facts_for_configmaps(self, namespace):
             name = item["metadata"]["name"]
             comp = self.comp(name)
             if comp is not None:
-                self.add_facts_for(comp, "configmaps", name, item["data"])
+                self.add_facts_for(comp, "configmaps", name, dict(item["data"]))
+                if comp not in ["fluentd"]:
```
Comment (richm) [inline roles/openshift_logging/library/openshift_logging_facts.py]: The throttle config is in yaml format

Comment (richm) [reply roles/openshift_logging/library/openshift_logging_facts.py]: will we need to update this if/when we provide the kibana config as a yaml formatted configmap?

File: roles/openshift_logging/filter_plugins/openshift_logging.py (hunk 5 above)
Comment (richm) [reply roles/openshift_logging/filter_plugins/openshift_logging.py]: What I mean is - does `line` need to be regex escaped?  bash shell escaped?

File: roles/openshift_logging/library/openshift_logging_facts.py (hunk 7 above)
Comment (ewolinetz) [reply roles/openshift_logging/library/openshift_logging_facts.py]: There may need to be a more elegant way to handle this then... I can't remember why I made a special case out of fluentd and I'm surprised there isn't an issue with the fact that  throttle config is in yaml format. It may just be because we're currently only pulling values out of the ES configmap...?

File: roles/openshift_logging/filter_plugins/openshift_logging.py (hunk 5 above)
Comment (ewolinetz) [reply roles/openshift_logging/filter_plugins/openshift_logging.py]: I added comments for the wrong method... this can likely be scrapped. We aren't using it, it's been functionally replaced by `account_for_whitelist` in `lib/logging_patch.py`

File: roles/openshift_logging/library/openshift_logging_facts.py [hunk 8]
```diff
@@ -214,7 +222,10 @@ def facts_for_configmaps(self, namespace):
             name = item["metadata"]["name"]
             comp = self.comp(name)
             if comp is not None:
-                self.add_facts_for(comp, "configmaps", name, item["data"])
+                self.add_facts_for(comp, "configmaps", name, dict(item["data"]))
+                if comp is in ["elasticsearch", "elasticsearch_ops"]:
```
Comment (richm) [inline roles/openshift_logging/library/openshift_logging_facts.py]: remove the `is` - `comp in [...`

Comment (ewolinetz) [reply roles/openshift_logging/library/openshift_logging_facts.py]: whoops

**Comment by richm** [comment]
```
FAILURE after 624.744s: test/fluentd-forward.sh:133: executing 'wait_for_fluentd_to_catch_up' expecting success: the command returned the wrong error code
```
flake?
/test logging

**Comment by ewolinetz** [comment]
@michaelgugino can i get an approval? :)
